Direct Associations (#9182)
mendix/docs: content/en/docs/refguide/modeling/domain-model/associations/association-storage.md

XPath and OQL queries work identically for both association tables and direct associations. You do not have to change anything or learn different flavors of these languages to work with them.

## Advantages of Direct Associations

Because they don't have association tables, using direct associations can bring the following advantages:

* In many cases, using direct associations can speed up queries and statements
* They use less space in the database
* You can choose to implement direct associations for some of your associations, while leaving others as association tables

## Limitations of Direct Associations

You cannot replace association tables with direct associations for all types of association in your app. Direct associations have the following limitations: 

* You can only use direct associations for one-to-many (1-*, default owner) and one-to-one (1-1, owner both) associations – you cannot use it for many-many associations.

> * You can only use direct associations between persistable and remote entities. You cannot use them on view entities or non-persistable entities.

Reviewer: I think it would be good to be a more explicit here:

> Direct associations are only allowed between persistable entities, between local entities and from a local to a remote entity. You cannot use them on view entities, non-persistable entities or between remote entities.

Author: Updated.

> {{% todo %}}Originally said "remote persistent" - I don't think we document two different types of remote entity?{{% /todo %}}

Reviewer: Im not that knowledgeable on remote entities, but the documentation mentions the ability to make non-persistant remote entities 
https://docs.mendix.com/refguide/external-entities/#general

Author: I've put it back in.

## Enabling Direct Associations

Enabling direct associations is simple, and has the following features:

* The choice is reversible – you can decide to revert to using association tables (but see things to think about [before switching to direct associations](#before))
* Direct associations are available in Mendix versions 10.21 and above
* You can decide whether to enable it for a specific association, or [make it the default](/brokenlink) for all new associations

For more information, see the [Association Storage](/refguide/association-properties/#storage) section of *Association Properties*.

> {{% todo %}}Do we want to say this without a firm timeline?{{% /todo %}}

Reviewer: I realize now that it would be better not to include planned features as a promise in the documentation. This is something i would like to hear pavels opinion on as well.

Author: @dahfjkg - do you agree?

Reviewer 2: IMO, users open docs primarily to understand existing functionality, so I agree with Pijus

Author: Removed.

The development pipeline includes plans to bulk migrate the database to avoid long downtime for large databases.

## Before Switching to Direct Associations{#before}

Before deciding to switch from an association table to a direct association, bear the following in mind:

* Associations have to be rewritten to the database so migration can take a long time, especially where you have a large amount of data already stored in your database 
* Queries are not always faster, and might not be faster in your use case

> * If you have written any custom SQL, this might break – existing XPaths and OQL queries will not be affected

Reviewer: Maybe to clarify: 
>If you have written any custom SQL that accesses mendix database tables directly....

---
title: "Association Storage Options"
linktitle: "Association Storage"
url: /refguide/association-storage/
weight: 25
---

## Introduction

Association storage options were introduced in Mendix 10.21 to give you more control over how associations are implemented in the underlying database.

## Association Tables vs. Direct Associations

<!--- Everything except simple associations is in tables – what do you mean "except simple associations?" --->

> Prior to Mendix 10.21, all associations were stored in association tables. This had the advantage that you didn't have to worry about the [multiplicity](/refguide/association-properties/#multiplicity) or [navigability](/refguide/association-properties/#navigability) of the associations. You could change things as your domain model evolved.

Reviewer 2: > Prior to Mendix 10.21, all associations were stored in association tables

That is not true for Owner and ChangedBy. Those can be used in XPaths, so I guess users perceive them as associations

Author: How about
```suggestion
Prior to Mendix 10.21, all associations (except `Owner` and `ChangedBy`, where selected) were stored in association tables. This had the advantage that you didn't have to worry about the [multiplicity](/refguide/association-properties/#multiplicity) or [navigability](/refguide/association-properties/#navigability) of the associations. You could change things as your domain model evolved.
```

Reviewer 3: The documentation sometimes calls them "system associations".  Also, they aren't capitalized. See https://docs.mendix.com/refguide/entities/#properties. So maybe "except the system associations `owner` ..."?

Author: Thanks for the review. Missed the proper references, sorry. I'll change it to
```suggestion
Prior to Mendix 10.21, all associations (except the system associations `owner` and `changedby`, where selected) were stored in association tables. This had the advantage that you didn't have to worry about the [multiplicity](/refguide/association-properties/#multiplicity) or [navigability](/refguide/association-properties/#navigability) of the associations. You could change things as your domain model evolved.
```

In Mendix 10.21 you can choose to implement some associations as direct associations. This means that the ID of the **Child** object is stored as a foreign key column of the **Parent** object (for example the "many" side of the association) in the underlying database table, thus removing the need for a association table.

XPath and OQL queries work identically for both association tables and direct associations. You do not have to change anything or learn different flavors of these languages to work with them.

## Default Association Storage

In Mendix 10.20 and below associations are always implemented as association tables.

In Mendix 10.21 and above, the following defaults apply:

* **New projects** – one-to-many and one-to-one associations are implemented as direct associations
* **Upgraded projects** – for projects which are upgraded from an older version of Mendix, all new associations continue to be implemented as association tables

{{% alert color="info" %}}

> In your app settings you can [change the default](/refguide/app-settings/#miscellaneous) for all new associations.

Reviewer 2: That is implied, but I would mention explicitly that the setting does not affect existing associations.

Author: Done

{{% /alert %}}

## Advantages of Direct Associations

Because they don't have association tables, using direct associations can bring the following advantages:

> * In many cases, using direct associations can speed up queries and statements

Reviewer 2: > queries and statements

This language is a bit technical, and there is no way to run a statement from Mendix (we are working on data modifications in OQL though). I would mention something closer to everyday Mendix development like data grids and data modifications.

Author: Good point. How about
```suggestion
* In many cases, using direct associations can speed up populating data grids and changing entities
```

Reviewer 2: Maybe like this?
```
In many cases, using direct associations can speed up populating data grids, changing entities and other actions that involve database interactions
```
Or even
```
In many cases, using direct associations can speed up data retrieval and modification
```

Author: I like the shorter version, so how about:
```suggestion
* In many cases, using direct associations can speed up data retrieval and modification
```

Reviewer 2: Looks good

* They use less space in the database
* You can choose to implement direct associations for some of your associations, while leaving others as association tables

## Limitations of Direct Associations

You cannot replace association tables with direct associations for all types of association in your app. Direct associations have the following limitations: 

* You can only use direct associations for one-to-many (1-*, default owner) and one-to-one (1-1, owner both) associations – you cannot use it for many-many associations.

> * Direct associations are only allowed between persistable entities or between a persistable entity and a persistable external entity. You cannot use them on view entities, non-persistable entities, or between external entities.

Reviewer 3: Not sure I'd call that a limitation. Those types of associations are not stored in the local database, so one cannot say that for these "you cannot replace association tables", as they don't have any in the first place.

Author: So miss out the second bit? I think we need to mention it somewhere and it is a limitation in that the option doesn't appear for other entities (albeit there is a very good reason for that).
```suggestion
* Direct associations are only allowed between persistable entities or between a persistable entity and a persistable external entity.
```

Reviewer 3: How about something like:

Association storage only applies to associations stored in the local database, those are associations between persistable entities or between a persistable entity and a persistable external entity. Associations from view entities, non-persistable entities or external entities are not stored in the local database, so they do not have an association storage option.

Author: ```suggestion
* Association storage only applies to associations stored in the local database, that is associations between persistable entities or between a persistable entity and a persistable external entity. Associations from view entities, non-persistable entities, or external entities are not stored in the local database, so they do not have an association storage option.
```

## Enabling Direct Associations

Enabling direct associations is simple, and has the following features:

* The choice is reversible – you can decide to revert to using association tables (but see things to think about [before switching to direct associations](#before))
* Direct associations are available in Mendix versions 10.21 and above

> * You can decide whether to enable it for a specific association, or [change the default](/refguide/app-settings/#miscellaneous) for all new associations

Reviewer 2: These options are not exclusive, so I would make them separate points

Author: How about:
```suggestion
* You can [change the default](/refguide/app-settings/#miscellaneous) for all new associations
* You can enable it for specific associations
```

For more information, see the [Association Storage](/refguide/association-properties/#storage) section of *Association Properties*.

{{% todo %}}Do we want to say this without a firm timeline?{{% /todo %}}

> The development pipeline includes plans to bulk migrate the database to avoid long downtime for large databases.

Reviewer 3: The only plan we have is to make it easier in Studio Pro to change the setting on many associations. This would just be a convenience feature to avoid design time clicks and would in no way reduce the deployment downtime.

Author: I've removed this as we don't normally document future improvements - although we can tell individual customers if they ask about it.

## Before Switching to Direct Associations{#before}

Before deciding to switch from an association table to a direct association, bear the following in mind:

* Associations have to be rewritten to the database so migration can take a long time, especially where you have a large amount of data already stored in your database 
* Queries are not always faster, and might not be faster in your use case

> * If you have written any custom SQL that accesses Mendix tables directly, this might break – existing XPaths and OQL queries will not be affected

Reviewer 2: Using " – " here is probably correct, but for me as a native Russian speaker it means that what follows adds details to what stands before it. As there are more of us working in software, let's use "but" instead 😅

Author: Done.
